### ISSUE TITLE
Allow patterns in lambda bindings?

### ISSUE THREAD

**Issue by byorgey** [disco-lang/disco#182]
For example, `(λ(x,y). x + y) : N*N -> N`?  It's probably not worth having syntax for multiple clauses though.  So this would only be useful for things like tuples.

**Comment by byorgey** [comment]
This is much more important now that things are uncurried.  So for example if you want to do a fold with a non-standard operator you would want to write something like
```
reduce(\(x,y). x+2y, 0, bs)
```
The alternative is to define a separate named helper function, or write an inline case, neither of which is very nice:
```
f (x,y) = x+2y
reduce(f, 0, bs)

reduce(\p. {? x+2y when p is (x,y) ?}, 0, bs)
```

**Comment by byorgey** [comment]
This requires
* Updating AST to allow patterns in lambda binders instead of just names (but only in surface and typed syntax)
* Updating parser to parse it
* Updating the typechecker
* Updating the desugarer to desugar pattern bindings into normal bindings + case.  In fact, we already do this for definitions (the difference being that with definitions there can be multiple clauses).  Should find a way to reuse the code.

**Comment by byorgey** [comment]
Closed by #198.
